src/test/scala/com/eclipsesource/schema/SchemaValidatorSpec.scala

```diff
@@ -165,39 +167,37 @@ class SchemaValidatorSpec extends PlaySpecification {
       }
     }
 
-  "should resolve references on the classpath via UrlResolver" in {
-    val validator = SchemaValidator().addUrlResolver(ClasspathUrlResolver())
+  "should resolve references on the classpath with ClasspathUrlResolver" in {
+    val validator = SchemaValidator().addUrlResolver(ClasspathUrlProtocolHandler())
     // some.json references location.json within JAR
     val someJson = getClass.getResourceAsStream("/some.json")
     val schema = JsonSource.schemaFromStream(someJson)
     val result = validator.validate(schema.get, Json.obj("location" -> Json.obj("name" -> "Munich")))
     result.isSuccess must beTrue
   }
 
-  "should resolve references on the classpath via UrlResolver and custom protocols (#65)" in {
+  "should resolve relative references on the classpath with relative URL handler (#65)" in {
     val validator = SchemaValidator()
-      .addUrlResolver(ClasspathUrlResolver())
-      .shouldResolveRelativeRefsWithCustomProtocols(true)
-    // some.json references location.json within JAR
+      .addRelativeUrlHandler(new ClasspathHandler)
+    // some-issue-65.json references location.json within JAR
     val someJson = getClass.getResourceAsStream("/some-issue-65.json")
     val schema = JsonSource.schemaFromStream(someJson)
     val result = validator.validate(schema.get, Json.obj("location" -> Json.obj("name" -> "Munich")))
     result.isSuccess must beTrue
   }
 
-  "should resolve relative references with custom protocols (#65)" in {
+  "should resolve relative references on the classpath with relative URL handler (#65 - 2)" in {
     val validator = SchemaValidator()
-      .addUrlResolver(ClasspathUrlResolver())
-      .shouldResolveRelativeRefsWithCustomProtocols(true)
+      .addRelativeUrlHandler(new ClasspathHandler)
     val mySchemaJson = getClass.getResourceAsStream("/my-schema.schema")
     val mySchema = JsonSource.schemaFromStream(mySchemaJson)
     val result = validator.validate(mySchema.get, Json.obj("foo" -> Json.obj("bar" -> "whatever")))
     result.isSuccess must beTrue
   }
 
-  "should fail resolving relative references with custom protocols if option not set(#65)" in {
+  "should fail resolving relative references without any relative URL handler available (#65)" in {
     val validator = SchemaValidator()
-      .addUrlResolver(ClasspathUrlResolver())
+      .addUrlResolver(ClasspathUrlProtocolHandler())
```

Comment:
I think I mostly agree with you. I would not necessarily say that it's up to the implementer to handle a relative path vs. non-relative. I may be misunderstanding your statement, but I think any handler should handle any path and not know/care whether it is relative. It just opens a connection to the URL provided (in essence, the path it receives must always be absolute). It would then be up to your engine to determine how to build a path to pass to the handler. If the handler were in the set of "absolute" handlers, it expects the path in the schema to be absolute and passes it with little to no change. If the handler were in a set of "relative" handlers, you engine expects relative paths in the schemas and builds the appropriate absolute path to pass to the handlers.

That way, from the handler perspective, it only has to know whether or not it is tied to a protocol, not whether or not is is relative. Protocols and relative-ness should be completely decoupled concepts.

I would say that protocol-ful handlers get tried first, so that they have precedence (absolute handlers, then relative). Then try the protocol-less ones in order until one is successful (similarly absolute handlers, then relative). I think it's reasonable to expect the user of your APIs to know that they are providing the handlers in the order that they will be tried.

Just a note: order-wise, I have not fully thought that through, but that is an initial concept/idea.